src/goto-instrument/contracts/dynamic-frames/dfcc_utils.cpp

```diff
@@ -0,0 +1,577 @@
+  const symbolt *old_sym = symbol_table.lookup(function_id);
+  symbolt wrapped_sym;
+  wrapped_sym = *old_sym;
+  wrapped_sym.name = wrapped_function_id;
+  wrapped_sym.base_name = wrapped_function_id;
+  wrapped_sym.location = sl;
+  const auto wrapped_found = symbol_table.insert(std::move(wrapped_sym));
+  INVARIANT(
+    wrapped_found.second,
+    "dfcc_utilst::wrap_function: wrapped function symbol " +
+      id2string(wrapped_function_id) + " already exists in the symbol table");
+
+  // Re-insert a symbol for `function_id` which is now the wrapper function
+  symbolt wrapper_sym;
+  wrapper_sym = *old_sym;
+
+  std::function<const irep_idt(const irep_idt &)> trans_param =
+    [&](const irep_idt &old_param) {
+      return id2string(old_param) + "_wrapper";
+    };
+
+  // create new code_type with renamed parameters for the wrapper
+  const auto &old_code_type = to_code_type(old_sym->type);
+  code_typet::parameterst new_params;
+  clone_parameters(
+    old_code_type.parameters(),
+    wrapper_sym.mode,
+    wrapper_sym.module,
+    wrapper_sym.location,
+    // the naming scheme is `function_id::param` + `param_suffix`
+    trans_param,
+    function_id,
+    new_params);
+
+  code_typet new_code_type(new_params, old_code_type.return_type());
+
+  wrapper_sym.type = new_code_type;
+
+  // Remove original symbol from the symbol_table
+  if(goto_model.symbol_table.remove(function_id))
+  {
+    log.error() << "dfcc_utilst::wrap_function: could not remove "
+                << function_id << " from the symbol table" << messaget::eom;
+    exit(0);
+  }
+
+  // Insert update symbol in the symbol_table
+  const auto wrapper_sym_found = symbol_table.insert(std::move(wrapper_sym));
+  INVARIANT(
+    wrapper_sym_found.second,
+    "dfcc_utilst::wrap_function: could not insert symbol " +
+      id2string(function_id) + " in the symbol table");
+
+  // Insert wrapper function in the function_map
+  goto_functiont &wrapper_fun = goto_functions.function_map[function_id];
+  wrapper_fun.set_parameter_identifiers(new_code_type);
+  wrapper_fun.body.add(goto_programt::make_end_function(sl));
+}
+
+const exprt dfcc_utilst::make_null_check_expr(const exprt &ptr)
+{
+  return equal_exprt(ptr, null_pointer_exprt(to_pointer_type(ptr.type())));
+}
+
+exprt dfcc_utilst::make_sizeof_expr(const exprt &expr)
+{
+  log.debug() << "dfcc_utilst::make_sizeof_expr(" << format(expr) << ")"
+              << messaget::eom;
+  const auto &size =
+    size_of_expr(expr.type(), namespacet(goto_model.symbol_table));
+
+  PRECONDITION_WITH_DIAGNOSTICS(
+    size.has_value(),
+    "dfcc_utilst::make_sizeof_expr: no definite size for lvalue target:\n" +
+      expr.pretty());
+
+  return typecast_exprt::conditional_cast(size.value(), size_type());
+}
+
+exprt dfcc_utilst::make_map_start_address(const exprt &expr)
+{
+  return typecast_exprt::conditional_cast(
+    address_of_exprt(index_exprt(expr, from_integer(0, c_index_type()))),
+    pointer_type(bool_typet()));
```

Comment:
Use `to_array_type(expr.type()).index_type()` instead of `c_index_type()`

Reply:
fixed